src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIMarsHandle.cs

```diff
@@ -0,0 +1,521 @@
+                            continue;
+                        }
+                        else
+                        {
+                            _ackEvent.Set();
+                        }
+                    }
+
+                    break;
+                }
+            }
+
+            return TdsEnums.SNI_SUCCESS;
+        }
+
+        /// <summary>
+        /// Send a packet asynchronously
+        /// </summary>
+        /// <param name="packet">SNI packet</param>
+        /// <param name="callback">Completion callback</param>
+        /// <returns>SNI error code</returns>
+        public override uint SendAsync(SNIPacket packet, SNIAsyncCallback callback = null)
+        {
+            lock (this)
+            {
+                _sendPacketQueue.Enqueue(new SNIMarsQueuedPacket(packet, callback != null ? callback : HandleSendComplete));
+            }
+
+            SendPendingPackets();
+            return TdsEnums.SNI_SUCCESS_IO_PENDING;
+        }
+
+        /// <summary>
+        /// Receive a packet asynchronously
+        /// </summary>
+        /// <param name="packet">SNI packet</param>
+        /// <returns>SNI error code</returns>
+        public override uint ReceiveAsync(ref SNIPacket packet)
+        {
+            lock (_receivedPacketQueue)
+            {
+                int queueCount = _receivedPacketQueue.Count;
+
+                if (_connectionError != null)
+                {
+                    SNILoadHandle.SingletonInstance.LastError = _connectionError;
+                    return TdsEnums.SNI_ERROR;
+                }
+
+                if (queueCount == 0)
+                {
+                    _asyncReceives++;
+                    return TdsEnums.SNI_SUCCESS_IO_PENDING;
+                }
+
+                packet = _receivedPacketQueue.Dequeue();
+
+                if (queueCount == 1)
+                {
+                    _packetEvent.Reset();
+                }
+            }
+
+            lock (this)
+            {
+                _receiveHighwater++;
+            }
+
+            SendAckIfNecessary();
+            return TdsEnums.SNI_SUCCESS;
+        }
+
+        /// <summary>
+        /// Handle receive error
+        /// </summary>
+        public void HandleReceiveError()
+        {
+            lock (_receivedPacketQueue)
+            {
+                _connectionError = SNILoadHandle.SingletonInstance.LastError;
+                _packetEvent.Set();
+            }
+#if MANAGED_SNI
+            ((TdsParserStateObject)_callbackObject).ReadAsyncCallback(null, 1);
+#endif
+        }
+
+        /// <summary>
+        /// Handle send completion
+        /// </summary>
+        /// <param name="packet">SNI packet</param>
+        /// <param name="sniErrorCode">SNI error code</param>
+        public void HandleSendComplete(SNIPacket packet, uint sniErrorCode)
+        {
+            lock (this)
+            {
+                Debug.Assert(_callbackObject != null);
+#if MANAGED_SNI
```

Comment:
This file appears to only be included when MANAGED_SNI is being set.  Are these #ifs needed?